opm/simulators/linalg/bda/BILU0.cpp

```diff
@@ -129,6 +153,233 @@ namespace bda
+        }
+
+        // for every row
+        for (int i = 0; i < Nb; i++) {
+            int iRowStart = LUmat->rowPointers[i];
+            int iRowEnd = LUmat->rowPointers[i + 1];
+            // for every block in this row
+            for (int ij = iRowStart; ij < iRowEnd; ij++) {
+                int j = LUmat->colIndices[ij];
+                if (i <= j) {
+                    Ut->rowPointers[j+1]++;   // actually colPointers
+                } else {
+                    Lmat->colIndices[num_blocks_L] = j;
+                    memcpy(Lmat->nnzValues + num_blocks_L * bs * bs, LUmat->nnzValues + ij * bs * bs, sizeof(double) * bs * bs);
+                    num_blocks_L++;
+                }
+            }
+            Lmat->rowPointers[i+1] = num_blocks_L;
+        }
+
+        // prefix sum
+        int sum = 0;
+        for (int i = 1; i < Nb+1; i++) {
+            sum += Ut->rowPointers[i];
+            Ut->rowPointers[i] = sum;
+        }
+
+        // for every row
+        for (int i = 0; i < Nb; i++) {
+            int iRowStart = LUmat->rowPointers[i];
+            int iRowEnd = LUmat->rowPointers[i + 1];
+            // for every block in this row
+            for (int ij = iRowStart; ij < iRowEnd; ij++) {
+                int j = LUmat->colIndices[ij];
+                if (i <= j){
+                    int idx = Ut->rowPointers[j]++;
+                    Ut->colIndices[idx] = i;     // actually rowIndices
+                    memcpy(Ut->nnzValues + idx * bs * bs, LUmat->nnzValues + ij * bs * bs, sizeof(double) * bs * bs);
+                }
+            }
+        }
+
+        // rotate
+        // the Ut->rowPointers were increased in the last loop
+        for (int i = Nb; i > 0; --i) {
+            Ut->rowPointers[i] = Ut->rowPointers[i-1];
+        }
+        Ut->rowPointers[0] = 0;
+
+        Opm::Detail::Inverter<bs> inverter;
+
+        // Utmp is needed for CPU and GPU decomposition, because U is transposed, and reversed at the end
+        // Ltmp is only needed for CPU decomposition, GPU creates GPU buffer for Ltmp
+        double *Utmp = new double[Ut->nnzbs * block_size * block_size];
+
+        // actual ILU decomposition
+#if CHOW_PATEL_GPU
+        fgpilu.decomposition(queue, context,
+                    Ut->rowPointers, Ut->colIndices, Ut->nnzValues, Ut->nnzbs,
+                    Lmat->rowPointers, Lmat->colIndices, Lmat->nnzValues, Lmat->nnzbs,
+                    LUmat->rowPointers, LUmat->colIndices, LUmat->nnzValues, LUmat->nnzbs,
+                    Nb, num_sweeps, verbosity);
+#else
+        double *Ltmp = new double[Lmat->nnzbs * block_size * block_size];
+        for (int sweep = 0; sweep < num_sweeps; ++sweep) {
+
+            // for every row
+            for (int row = 0; row < Nb; row++) {
+                // update U
+                int jColStart = Ut->rowPointers[row];
+                int jColEnd = Ut->rowPointers[row + 1];
+                // for every block in this row
```

Comment:
I might be missing something again, but:
If we would use U instead of Ut here and do the transpose later, then we do not even need to search for U_ij as counting from the diagonal would suffice.

Reply:
It is quite difficult, I added some comments in 123e3fa.
I'm now testing the approach below. It looks like the two `idxL` loops are the same, except for their max value.
```
// for every row in A
for(int i = 0; i < Nb; ++i){
    int rowStart = LUmat->rowPointers[i];
    int rowEnd = LUmat->rowPointers[i+1];
    // for every block Aij on this row
    for(int ij = rowStart; ij < rowEnd; ++ij){
        int j = LUmat->colIndices[ij];
        if (i > j) {
            // calculate Lij
            // for every block on row i of L with col k (use A but stop at j)
            // try to find block on col j of U with row k

            // for every block on row i of L up to col j (block ij has col j)
            for(int idxL = rowStart; idxL < ij; ++idxL){
                int k = LUmat->colIndices[idxL];
                // on row k of U, try to find block with col j
                // start looking from diagonal block on row k
                int idxU = diagIndex[k];
                int colU = LUmat->colIndices[idxU];
                while(colU < j){
                    idxU++; // move to next block on rowU
                    colU = LUmat->colIndices[idxU];
                }
                if(colU == j){
                    // Aij -= (Lik * Ukj)
                }
            }
        } else {
            // calculate Uij
            // for every block on row i of L with col k (use A but stop at diagonal)
            // try to find block on col j of U with row k

            // for every block on row i of L up to diag
            for(int idxL = rowStart; idxL < diag[i]; ++idxL){
                int k = LUmat->colIndices[idxL];
                // on row k of U, try to find block with col j
                // start looking from diagonal block on row k
                int idxU = diagIndex[k];
                int colU = LUmat->colIndices[idxU];
                while(colU < j){
                    idxU++; // move to next block on rowU
                    colU = LUmat->colIndices[idxU];
                }
                if(colU == j){
                    // Aij -= (Lik * Ukj)
                }
            }
        }
    }
}```

Reply:
Will you try more or should we skip this?